src/sync/app.cpp

```
@@ -0,0 +1,153 @@
+
+    auto handler = [&](const Response response) {
+        // if there is a already an error code, pass the error upstream
+        if (response_code_is_fatal(response.status_code)) { // FIXME: handle
+            return completion_block(nullptr, handle_error(response));
+        }
+
+        auto j = nlohmann::json::parse(response.body);
+
+        realm::SyncUserIdentifier identifier {
+            j["user_id"].get<std::string>(),
+            m_auth_route
+        };
+
+        std::shared_ptr<realm::SyncUser> sync_user;
+        try {
+            sync_user = realm::SyncManager::shared().get_user(identifier,
+                                                              j["refresh_token"].get<std::string>(),
+                                                              j["access_token"].get<std::string>());
+        } catch (const error::AppError& err) {
+            return completion_block(nullptr, std::make_unique<error::AppError>(err));
+        }
+
+        std::string profile_route = util::format("%1/auth/profile", m_base_route);
+        std::string bearer = util::format("Bearer %1", sync_user->access_token());
+
+        GenericNetworkTransport::get()->send_request_to_server({
+            Method::get,
+            profile_route,
+            timeout_ms,
+            {
+                { "Content-Type", "application/json;charset=utf-8" },
+                { "Accept", "application/json" },
+                { "Authorization", bearer}
+            },
+            std::string()
+        }, [&](const Response profile_response) {
+            // if there is a already an error code, pass the error upstream
+            if (response_code_is_fatal(profile_response.status_code)) {
+                return completion_block(nullptr, handle_error(profile_response));
+            }
+
+            j = nlohmann::json::parse(profile_response.body);
```

If there is a bug or change in the json received, this will crash the caller hard. We should probably wrap most of the logic here in a try-catch so we can return an appropriate error if there is a programmer/server error somewhere we didn't expect.